conmon-rs/server/src/child_reaper.rs

```diff
@@ -0,0 +1,156 @@
+                    Ok(status) => {
+                        if let WaitStatus::Exited(grandchild_pid, exit_status) = status {
+                            let grandchildren = Arc::clone(&grandchildren);
+                            thread::spawn(move || {
+                                if let Err(e) = Self::forget_grandchild(
+                                    grandchildren,
+                                    grandchild_pid,
+                                    exit_status,
+                                ) {
+                                    error!(
+                                        "Failed to reap grandchild for pid {}: {}",
+                                        grandchild_pid, e
+                                    );
+                                }
+                            });
+                        }
+                    }
+
+                    Err(err) => {
+                        // TODO FIXME this busy loops right now while there are no grandchildren.
+                        // There should be a broadcast mechanism so we only run this loop
+                        // while there are grandchildren
+                        if err != nix::errno::Errno::ECHILD {
+                            error!("caught error in reading for sigchld {}", err);
+                        }
+                    }
+                }
+            }
+        });
+        Ok(())
+    }
+
+    pub fn create_child<P, I, S>(
+        &self,
+        cmd: P,
+        args: I,
+    ) -> Result<std::process::ExitStatus, std::io::Error>
+    where
+        P: AsRef<std::ffi::OsStr>,
+        I: IntoIterator<Item = S>,
+        S: AsRef<std::ffi::OsStr>,
+    {
+        let mut cmd = std::process::Command::new(cmd);
+        cmd.args(args);
+
+        // To prevent an error: "No child processes (os error 10)",
+        // we prevent our reaper thread from calling waitpid until this child has completed.
+        // Inspired by
+        // https://github.com/kata-containers/kata-containers/blob/828a304883a4/src/agent/src/signal.rs#L27
+        let wait_lock = Arc::clone(&self.wait_lock);
+        let _lock = wait_lock.lock().unwrap();
+
+        cmd.spawn()?.wait()
+    }
+
+    pub fn watch_grandchild(&self, id: String, pid: i32, exit_paths: Vec<PathBuf>) -> Result<()> {
+        let locked_grandchildren = Arc::clone(&self.grandchildren);
+        let mut map = locked_grandchildren.lock().unwrap();
+
+        let reapable_grandchild = ReapableChild { id, exit_paths };
+        if let Some(old) = map.insert(pid, reapable_grandchild) {
+            bail!("Repeat PID for container {} found", old.id);
+        }
+        Ok(())
+    }
+
+    fn forget_grandchild(
+        locked_map: Arc<Mutex<HashMap<i32, ReapableChild>>>,
+        grandchild_pid: Pid,
+        exit_status: i32,
+    ) -> Result<()> {
+        debug!("caught signal for pid {}", grandchild_pid);
+
+        let mut map = locked_map.lock().unwrap();
```

Comment:
Let's do not unwrap:
```suggestion
        let mut map = locked_grandchildren
            .lock()
            .map_err(|e| format_err!("lock grandchildren: {}", e))?;
```